ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+    MempoolCapacityBytes
+  $ getMempoolCapacityBytes modelCapacity
+  - sum [ txInBlockSize tx
+        | (tx, _) <- modelTxs
+        ]
+
+-- | Actions on the mempool
+data Action blk r =
+    -- | Add some transactions to the mempool
+    NewTxs ![GenTx blk]
+  | -- | Unconditionally sync with the ledger db
+    SyncLedger
+  | -- | Ask for the current snapshot
+    GetSnapshot
+  -- maybe add ? | GetSnapshotFor (Point blk)
+  deriving (Generic1)
+  deriving (Rank2.Functor, Rank2.Foldable, Rank2.Traversable, CommandNames)
+
+-- | Events external to the mempool
+data Event blk r = ChangeLedger
+    !(LedgerState blk ValuesMK)
+    !Bool -- ^ Is this a new fork? i.e. should the current mempool tip became
+          -- unreachable? TODO use sth isomorphic with Bool
+  deriving (Generic1)
+  deriving (Rank2.Functor, Rank2.Foldable, Rank2.Traversable, CommandNames)
+
+data Command blk r = Action !(Action blk r)
+                   | Event  !(Event blk r)
+  deriving (Generic1)
+  deriving (Rank2.Functor, Rank2.Foldable, Rank2.Traversable)
+
+instance CommandNames (Command blk) where
+  cmdName (Action action) = cmdName action
+  cmdName (Event event)   = cmdName event
+
+  cmdNames :: forall r. Proxy (Command blk r) -> [String]
+  cmdNames _ = cmdNames (Proxy @(Action blk r))
+            ++ cmdNames (Proxy @(Event blk r))
+
+data Response blk r =
+    -- | Nothing to tell
+    Void
+  | -- | Return the contents of a snapshot
+    GotSnapshot !(Snap blk)
+  deriving (Generic1)
+  deriving (Rank2.Functor, Rank2.Foldable, Rank2.Traversable)
+
+{-------------------------------------------------------------------------------
+  Model side
+-------------------------------------------------------------------------------}
+
+initModel ::
+  ( LedgerSupportsMempool blk
+  , ValidateEnvelope blk
+  )
+  => LedgerConfig blk
+  -> LedgerState blk ValuesMK
+  -> Model blk r
+initModel cfg initialState =
+  Model {
+    modelMempoolTip       = (ticked, True)
+  , modelLedgerDBTip      = initialState
+  , modelTxs              = []
+  , modelCapacity         = MempoolCapacityBytes 400
+  , modelLastTicket       = zeroTicketNo
+  , modelSeenStates       = Set.singleton $ castPoint $ getTip initialState
+  , modelMempoolWasFilled = False
+  }
+  where ticked = tick cfg initialState
+
+generator ::
+  ( Arbitrary (LedgerState blk ValuesMK)
+  , GetTip (LedgerState blk ValuesMK)
+  , StandardHash blk
+  , UnTick blk
+  )
+  => (Int -> LedgerState blk ValuesMK -> Gen [GenTx blk])
+     -- ^ Transaction generator based on an state
+  -> Model blk Symbolic
+  -> Maybe (Gen (Command blk Symbolic))
+generator gTxs Model{modelMempoolTip, modelSeenStates} =
+   Just $ sized $ \n ->
+    frequency
+      [ (100, Action . NewTxs <$> (gTxs n . unTick . fst $ modelMempoolTip))
+      , (10, pure $ Action SyncLedger)
+      , (10, do
+            b  <- arbitrary
+            ls <- arbitrary `suchThat` (  (`notElem` modelSeenStates)
```

Comment:
Now that I think about it, we might not even want this: don't we want to test situations in which the ledger db rolls back to a previous state?

Reply:
The problem is that it moves to the same `Point` and a different `UTxO` which is nonsense. This way we move to a different state.

I could instead do sth like:
```
arbitrary `suchThat` (\x -> 
  if x `notMember` modelSeenStates
  then getPoint x `notMember` map getPoint modelSeenStates 
  else True)
```

i.e. we can reuse states, but we cannot use an old point with a different utxo set.